File: lib_algorithms_sort/src/main/java/net/chris/lib/algorithms/sort/BubbleSorter.java
package net.chris.lib.algorithms.sort;

/**
 * Bubble sort.
 *
 * @param <T>
 */
public abstract class BubbleSorter<T> extends Sorter<T> {

    @Override
    protected void subSort(T[] array) {
        if (array == null) {
            return;
        }
        for (int i = 0; i < array.length - 1; i++) {
            for (int j = array.length - 1; j >= i + 1; j--) {
                T temp;
                if (compareTo(array[j], array[j - 1]) < 0) {
                    temp = array[j];
                    array[j] = array[j - 1];
                    array[j - 1] = temp;
                    doUpdate(array);
                }
            }
        }
    }

}

File: lib_algorithms_sort/src/main/cpp/BubbleSorter.cpp
#include "BubbleSorter.h"

inline void swap(int *array, int j, int k) {
    int temp;
    temp = array[j];
    array[j] = array[k];
    array[k] = temp;
}


void bubbleSort(int *array, int size) {
    for (int i = 1; i < size; i++) {
        for (int j = size - 1; j >= i; j--) {
            if (array[j] < array[j - 1]) {
                swap(array, j - 1, j);
            }
        }
    }
}